distributed/shuffle/_shuffle_extension.py

```diff
@@ -455,9 +523,21 @@ async def _get_shuffle(
                 return self.shuffles[shuffle_id]
 
     async def close(self) -> None:
+        if self.closed:
+            await self._closed_event.wait()
+            return
+
+        self.closed = True
         while self.shuffles:
             _, shuffle = self.shuffles.popitem()
             await shuffle.close()
+        self._closed_event.set()
+
+    def raise_if_closed(self) -> None:
+        if self.closed:
+            raise ShuffleClosedError(
+                f"{self.__class__.__name__} already closed on {self.worker.address}"
+            )
 
```

Comment:
IIUC this condition can only trigger if the worker is closing which closes the extension. This is beyond the scope of this PR but this problem strikes me as a fault of our server architecture. I don't think a server should accept any such handlers once it is closing.

However, the worker only closes the base class (listeners and handle_comms) at the very end, after plugins/extensions/etc. where closed.

I'm wondering two things right now
1. Why is this even necessary? What is happening if we just do not raise here? Worst case I can come up with is that the worker threads continue to do stuff but how bad is this really?
2. If that is an error we need to handle, we might want to change the exception message to a "worker closing" foo. From a users perspective there is no such thing as a "closed shuffle"

Reply:
Not raising anything here will fail one test `test_closed_worker_during_transfer`. The observed behavior of the test is still the same but the exception message is different

`RuntimeError: XY already closed` vs `RuntimeError: Trying to write data to closed buffer`

I'm wondering if instead of `raise_if_closed` everywhere we'd rather want a `try:...except` that reraises all exception to appropriate user exceptions instead

Reply:
If it's just about reraising, how about a decorator for all the comm handlers

```python
def _handle_exception(ext, handler):
    @functools.wraps(handler)
    def _(*args, **kwargs):
        try:
            return handler(*args, **kwargs)
        except Exception:
            if ext.closed:
                raise RuntimeError("Shuffle closed...")
            else:
                raise

    return _()
```